resource: load resources from config or R, and rework topo discovery

garlick opened the pull request:
This PR takes a step towards the architecture proposed in #3238.

The "resource object is still `hwloc.by_rank` and the `resource.acquire` RPC is unchanged in this PR.

A TOML config option is added to read the resource object from a file, e.g.:
```toml
[resource]
path = "/etc/flux/resource.json"
```

Code is added to fetch `R` from the enclosing instance, although conversion to a usable resource object is stubbed out for now.

If the resource object is known by one of these static methods, or because `resource.hwloc.by_rank` exists in the KVS when the resource module is loaded, then each broker validates its local shard of the resource object using hwloc (directly, not calling out to `flux hwloc`).  If validation fails, currently an error is logged and the resource module fails to load.  This needs to be changed to drain the node before this is merged.

If the resource object could not be determined up front, then resources are "discovered" as before, except the resource module now reduces the shards to a full object itself, rather than calling out to `flux hwloc reload`.  This may avoid some entanglements when `flux hwloc` is not finished when the instance exits as noted in #3235.

Resources can be faked by pre-populating `resource.hwloc.by_rank` in the KVS and loading the resource module with the `noverify` option.  It is no longer necessary to post an event to make this work.

One side effect of not using `flux hwloc reload` is that `resource.hwloc.xml`  is not populated.  However, a `resource.topo-get-xml` was added to allow it to be fetched directly from each rank on demand.  (Fluxion currently fetches each rank synchronously from the KVS in a loop so it should be relatively easy to replace the KVS lookups with this RPC).[1]

This is a WIP because it needs more tests, and it's still under discussion whether conversion to Rv1 should happen before or after this PR.

Edit: [1] Fluxion waits to fetch XML until `resource.acquire` returns.  When resources are discovered, that implies that all ranks are up.  However when resources are configured, acquire can return before all ranks are up, so this workaround for Fluxion may need more thought.

garlick commented:
OK, just forced a push after some refactoring and a rebase on current master.

All the `by_rank` specific stuff should now be localized in `byrank.[ch]`.

The hwloc specific stuff is relegated to `rhwloc.[ch]` which was plucked from the future (@grondo's forthcoming librset work).  One function was added to get xml from a `hwloc_topology_t`.

Resources are now read/written from `resource.R` in the KVS rather than the `by_rank` specific `resource.R`.

hwloc xml is collected at rank 0 and written to the usual place at `resource.hwloc.xml.<rank>`.  If resources are dynamically discovered, this is guaranteed to be in place by the time the first `resource.acquire` response is sent.  It will not be present if resources were configured from TOML or grabbed from `R` in the enclosing instance.  Hopefully this provides a bridge for Fluxion in the situations where we are using it now.

This should be relatively straightforward now to convert to Rv1.  Find the `byrank_*` function calls and replace that code, drop `rhwloc`.  This PR doesn't do anything with the `acquire` interface to move towards the proposed RFC 28 changes, but the `byrank` specific bits there are minimal.  For the most part the "resource object" is passed through without modification, except for  some code that calls `byrank_resobj_sub()` to remove ranks that are excluded from the initial response.

Which reminded me - Fluxion _does_ look at the resource object to excract the initial idset of ranks for a _grow_ operation (which it then satisfies with the hwloc XML data).  Uh on.

garlick commented:
Forced a push, adding `flux resource reload [--xml] [--force] PATH`.

This can be used to load test resources without reloading the resource module.

`PATH` can point to a file containing the JSON resource object, or if `--xml` is specified, a directory containing `<rank>.xml` files.  If the latter, the resource object is derived from the hwloc XML.

The resource object may not contain ranks that are not valid in the current instance (e.g. >= size), unless `--force` is used.

Resources are not validated against the hwloc topology when loaded in this way.

The scheduler still needs to be unloaded so the sequence used in test is
```
$ flux module remove scheduler
$ flux resource reload [args]
$ flux module load scheduler
```

dongahn commented:
> Which reminded me - Fluxion does look at the resource object to extract the initial idset of ranks for a grow operation (which it then satisfies with the hwloc XML data). Uh on.

Yes, this is true. Fluxion does "not" use the overall resource object but it does need to fetch the idset.

dongahn commented:
@garlick: If there are the things that Fluxion does now slow down this conversion, you shouldn't shy away calling it out even if that require some substantial changes.

dongahn commented:
> Edit: [1] Fluxion waits to fetch XML until resource.acquire returns. When resources are discovered, that implies that all ranks are up. However when resources are configured, acquire can return before all ranks are up, so this workaround for Fluxion may need more thought.

Is this still a problem? So the `acquire` will return and hwloc xml files (or RPC equivalent) are not there?

garlick commented:
>> Edit: [1] Fluxion waits to fetch XML until resource.acquire returns. When resources are discovered, that implies that all ranks are up. However when resources are configured, acquire can return before all ranks are up, so this workaround for Fluxion may need more thought.

> Is this still a problem? So the acquire will return and hwloc xml files (or RPC equivalent) are not there?

Since the above description, XML is now collected by the resource module and stored in the KVS, as before.  There are some small modifications to Fluxion proposed in flux-framework/flux-sched#766 to track the changes in this PR.  One change is to not load XML with KVS_FLAG_WAITCREATE, so if the XML is missing, it's a hard failure not a deadlock.

For the system instance (case 1)  in the short short term, we can initially not create a resource configuration, and boot the system instance with all nodes up once.  This will populate XML in the KVS.  Then on subsequent restarts, it will be reloaded from there and down nodes can be tolerated.

We can improve upon that by embedding JGF in the opaque scheduler section of a resource configuration and bootstrap the system instance from that.  Fluxion would need support for helping generate the resource configuration, and for bootstrapping from the JGF it gets back from `resource.acquire`.

For a sub-job (case 2) we need to discuss options - see https://github.com/flux-framework/flux-core/issues/3228#issuecomment-712305545.  In summary: The XML will not be there as of now.   It would be nice if minimally there was a capability for Fluxion to bootstrap from Rv1 alone.  We could also perhaps provide a hint to the resource module that it should collect XML in this case.  Passing through JGF and reranking seems hard, and might not be worth it in Rv1  time frame (best to move on to Rv2).

garlick commented:
Just pushed a change to drain nodes on R verification failure, and related tests.

garlick commented:
Just got this failure in travis. I was not expecting this to be racy, but possibly the async disconnect message from unloading sched-simple is not getting there before resource.acquire when it's being loaded?  Hmm, if that's going to be a problem in test then we might want to use an expicit cancel.
```
sched-simple: generate jobspec for simple test job

expecting success: 
	flux module unload sched-simple &&
	echo $hwloc_by_rank >R
	flux resource reload R &&
	flux module load sched-simple &&
	flux dmesg 2>&1 >reload.dmesg.log &&
	grep "ready:.*rank\[0-1\]/core\[0-1\]" reload.dmesg.log &&
	test_debug "echo result=\"$($query)\"" &&
	test "$($query)" = "rank[0-1]/core[0-1]"

2020-10-22T20:30:21.522533Z sched-simple.debug[0]: ready: 4 of 4 cores: rank[0-1]/core[0-1]
flux-resource: ERROR: [Errno 16] resources are busy (unload scheduler?)
```

garlick commented:
This was rebased on top of #3276, and everything converted to Rv1.

garlick commented:
Force pushed after a rebase on @grondo's latest version of #3276, and some minor tweaks here and there.

garlick commented:
Hmm, new failing test in travis only for `t2700-mini-cmd.t`:
```
expecting success: 
	jobid=$(flux mini submit --ntasks=2 --cores-per-task=2 hostname) &&
	flux job attach $jobid

0.024s: job.exception type=alloc severity=0 unsatisfiable request
not ok 6 - flux mini submit --ntasks=2 --cores-per-task=2 works
```
Couldn't reproduce by running under `taskset -c 0`.

grondo commented:
> Couldn't reproduce by running under taskset -c 0.

I think that particular test will be skipped if there are < 2 cores:
```
test $(nproc) -gt 1 && test_set_prereq HAVE_MULTICORE

[snip]

test_expect_success HAVE_MULTICORE 'flux mini submit --ntasks=2 --cores-per-task=2 works' '
        flux resource list &&
        jobid=$(flux mini submit --ntasks=2 --cores-per-task=2 hostname) &&
        flux job attach $jobid
'
```

I wonder if, in Travis, `nproc` returns a number > 1, but the processors are actual HTs and there is really only one core per rank available, so this job fails. This may be failing due to the fact we had to remove the "sched-pus" option from sched-simple.

garlick commented:
Rebased on @grondo's latest #3276 and added an RPC to fetch hwloc XML for the entire instance in one go.  XML is no longer stored in the KVS, and for now it is unconditionally collected by the resource module so that Fluxion can request it as needed.  If collection is still in progress, this RPC will delay the response until it completes.

I just pushed an update to flux-framework/flux-sched#764 that gets the "grow set" from Rv1 and pulls XML from the resource module directly.

Because the hwloc XML is not stored in the KVS, and all ranks are required to be online to collect it, Fluxion will require all ranks to be online in order to start scheduling.  This will be an inconvenience when using Fluxion as a system instance.  If possible, I would propose that we try to get Fluxion booting from Rv1 with embedded JGF during the month of Nov.  We can just deal with that limitation on fluke during the month of Nov, or we can discuss whether it makes sense to use sched-simple for our initial fluke testing.

garlick commented:
Restarted a bionic builder that failed in valgrind with:
```
2020-10-28T19:50:54.306730Z broker.err[0]: rc1.0: flux: flux_open: No such file or directory
68952020-10-28T19:50:54.374768Z broker.err[0]: rc1.0: /bin/bash -c /usr/src/etc/rc1 Exited (rc=1) 0.8s\
```

garlick commented:
Pushed one more time after another rebase on #3276.  I also increased the rc1 timeout on first open of the local connector in case that resolved the spurious valgrind failures.

garlick commented:
OK, fluxion tests are passing and travis seems good, so I think this could use a bit of review.

garlick commented:
Rebased on @grondo's latest, and added code to the resource module to look up `config.hostlist` (requested in #3283).  If defined, and R is supplied by configuration, then call `rlist_rerank()` to force the R mapping of ranks to hosts to match the broker's.

(Without this the sys admins would have to make sure the rank to host mapping in the [bootstrap] config and R were in sync, which would be tedious and error prone).

garlick commented:
Rebased on master and added a few more tests.

garlick commented:
And rebased again on #3283 (with test fix tacked on).

This time I noticed I had forgotten to squash out a large chunk of incremental development in by_rank.[ch].  Gone now.

garlick commented:
Restarted one builder that failed here
```
flux-start: 3 (pid 26801) Killed
5864flux-start: 1 (pid 26793) Killed
5865flux-start: 0 (pid 26792) Killed
5866not ok 4 - quorum reached on instance with 3 TBON levels
```

chu11 commented on src/modules/resource/inventory.c:
```diff
@@ -0,0 +1,652 @@
+ * Then commit the result to resource.R in the KVS, and post
+ * resource-define event to resource.eventlog.
+ *
+ * Alternatively, resource.R is pre-loaded into the KVS.
+ * If found, this takes precedence over case 2 but not 1.  This allows
+ * the R product to be reused upon module reload (saving work),
+ * or for resources to be faked in test.
```
could just list this as case 2 above, then case 2 above becomes case 3?

garlick replied:
well I was trying to stay synced with cases 1,2,3 as described in #3238 as that kind of became a shorthand in discussions.  The whole block comment is a bit lacking in detail though.  Let me work it over.

chu11 reviewed (commented):
skimming things, just a few small things

garlick commented on src/modules/resource/rutil.c:
```diff
@@ -236,6 +241,131 @@ bool rutil_match_request_sender (const flux_msg_t *msg1,
+
+    if (!(o = json_load_file (path, 0, &e))) {
+        snprintf (errbuf, errbufsize, "%s:%d %s", e.source, e.line, e.text);
+        if (errno == 0) {
```
oh yeah, that's just wrong.  I should set errno to ENOENT if access() fails and EPROTO otherwise.

garlick commented:
Rebased on current master and made a few fixes called out by @chu11.

There is now a way longer explanatory block comment at the top of `resource/inventory.c` that might need a little proof reading to make sure it makes sense and doesn't any typos that I missed (I've been so bad lately!)

chu11 commented:
> There is now a way longer explanatory block comment at the top of resource/inventory.c that might need a little proof reading to make sure it makes sense and doesn't include any typos that I missed (I've been so bad lately!)

Looks good to me now!

garlick commented:
Thanks! I actually have a few grammar changes to push to that big block comment. I'll go ahead and do that.

garlick commented:
Restarted centos 7 builder that failed here:
```
PASS: t0015-cron.t 31 - flux module remove cron
# passed all 31 test(s)
flux-start: 1 (pid 29202) Killed
flux-start: 0 (pid 29201) Killed
1..31
ERROR: t0015-cron.t - exited with status 137 (terminated by signal 9?)
```

grondo commented:
I just pushed some commits on top of this PR with some fixes and additional tests developed while running this branch through manual testing. @garlick, feel free to squash the incremental testing updates and small fix into previous commits if you approve of those changes.

garlick commented:
Those changes all LGTM.  Thanks!

grondo commented:
Ok, I've rebased on current master and squashed all fixup comments and forced a push to your branch @garlick.

grondo commented:
I had to force push again to fix an error in the new `get-xml-test.py` script caught by LGTM.

grondo reviewed (approved):
I gave this PR some extensive testing and I've contributed a few extra tests as a result, so now I'll approve. Since I've touched the PR maybe @chu11 could have another look as well if he has time.

grondo commented:
I restarted the flux-sched test since it appeared to be hanging

chu11 commented on src/modules/resource/monitor.c:
```diff
@@ -280,6 +285,84 @@ static int monitor_reduce_self (struct monitor *monitor, bool up)
+
+    if (flux_request_unpack (msg, NULL, "{s:i}", "up", &up) < 0) {
+        if (flux_respond_error (h, msg, errno, NULL) < 0)
+            flux_log_error (h, "error resonding to monitor-waitup request");
```
typo `resonding`

chu11 commented on src/modules/resource/monitor.c:
```diff
@@ -280,6 +285,84 @@ static int monitor_reduce_self (struct monitor *monitor, bool up)
+    }
+    if (up == count) {
+        if (flux_respond (h, msg, NULL) < 0)
+            flux_log_error (h, "error resonding to monitor-waitup request");
```
typo `resonding`

chu11 commented on src/modules/resource/monitor.c:
```diff
@@ -280,6 +285,84 @@ static int monitor_reduce_self (struct monitor *monitor, bool up)
+        msg = zlist_first (monitor->waiters);
+        while (msg) {
+            if (notify_one_waiter (monitor->ctx->h, count, msg)) {
+                zlist_remove (monitor->waiters, (void *)msg);
```
i don't think `zlist_remove()` is safe while iterating `zlist`.

grondo replied:
I asked about this as well, note that the function returns if any item is removed from the list, so this is safe. (I missed that as well, maybe a comment would help?)

chu11 replied:
Ahh, i see.  Yeah,  a comment would be good.

chu11 reviewed (commented):
did a not-super-thorough skim and caught a few things, nothing big.

grondo commented on src/common/librlist/rlist.c:
```diff
@@ -950,8 +950,11 @@ struct rlist *rlist_from_json (json_t *o, json_error_t *errp)
-    if (version != 1)
+    if (version != 1) {
+        if (errp)
+            sprintf (errp->text, "invalid version=%d", version);
```
the buffer is known to be smaller than the maximum possible error string (even if version is 12 figures)

grondo replied:
Although, you're right, why not use snprintf? I'll update.

grondo commented:
@garlick, I'll go ahead and force-push squashed fixups for the items @chu11 identified above.

Thanks @chu11!

grondo commented:
Ok, I've force-pushed an update addressing @chu11's comments.

garlick commented:
Nice, thanks guys!

> I restarted the flux-sched test since it appeared to be hanging

I'm not sure if I noticed whether we had hangs or not before, but the flux-sched test should not be expected to pass until we merge the flux-framework/flux-sched#766.

Think this is ready for MWP?

grondo commented:
Yeah, though MWP won't work since flux-sched checks won't pass. This will need a manual merge after removal of branch protections.

grondo commented:
I suppose I can go ahead and do that.

grondo commented:
Ok, I removed branch protections and manually merged. Huzzah!

garlick commented:
Yay!!!

On Tue, Nov 3, 2020, 7:54 PM Mark Grondona <notifications@github.com> wrote:

> Ok, I removed branch protections and manually merged. Huzzah!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/3265#issuecomment-721497704>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABJPW6BABT37JMLZZQUY7DSODGBFANCNFSM4SPINLLA>
> .
>
